src/client/Client.h

```diff
@@ -1329,14 +1329,6 @@ class Client : public Dispatcher, public md_config_obs_t {
     void finish(int r) override {
       CRF->finish_io(r);
     }
-
-    // For _read_async, we may not finish in one go, so be prepared for multiple
-    // calls to complete. All the handling though is in C_Read_Finisher.
-    void complete(int r) override {
-      finish(r);
-      if (CRF->iofinished)
-        delete this;
-    }
```

Comment:
This structure of contexts is mystifying (why is this wrapper even necessary??) but I guess this change is okay.

This code really needs tidying.

Reply:
Since this will inherit the parent class (`Context`) complete(), that would deallocate the object, so how is this supposed to work with multiple calls to the context callback?

Or, is there plan to fix that?

Reply:
Noted. I'm going to look at this closely and now that we there is effort to run `fio` tests with nfs-ganesha (w/ object cacher disabled), we have exercise these code paths in fs suite.

Reply:
@vshankar Could you please give reference on how to disable object-cacher? I would also like to run some `fio` tests with this fix on my samba-cephfs integration branch.

Reply:
> @vshankar Could you please give reference on how to disable object-cacher? I would also like to run some `fio` tests with this fix on my samba-cephfs integration branch.

@synarete it's a flag (`ceph/src/common/options/mds-client.yaml.in`) that can be set to `false`. If you run samba with vstart cluster then there is a bin called `ceph_test_client` that can be used with `--client_oc=false` or with teuthology i guess we now have `qa/workunits/client/test_oc_disabled.sh` (added in https://github.com/ceph/ceph/pull/59987) to test with your branch

Reply:
> Hmm, I need to look some stuff over. I think there's cruft from some intermediate stage of my development of non-blocking I/O that I didn't clean up. That was a very extended development process and I was stumbling over some very confusing stuff with Context.
> 
> Some of the wrappers are because we need an async Context for non-blocking but internally the I/O methods use their own Context.
> 

async i/o method has its own context:
```
  int64_t ll_preadv_pwritev(struct Fh *fh, const struct iovec *iov, int iovcnt,
                            int64_t offset, bool write,
                            Context *onfinish = nullptr,
                            bufferlist *blp = nullptr,
                            bool do_fsync = false, bool syncdataonly = false);
```
I guess it is totally fine to remove this override. The method's context should take care of it.

> But there seems to be an extra layer of complexity that may not actually be necessary. And what I am looking at right now, it looks like some stuff really is leaked...

Reply:
I think originally finish_io was going to have the possibility of returning without marking iofinished for a short read case. If that needs to be handled, then there's some work to do there, otherwise if it really isn't necessary, there's definitely some code simplification, we don't need iofinished with the current code...

The read code is incredibly complicated (and for awhile when I was calling the effort async instead of non-blocking mirroring Ganesha, we had async_sync instead of nonblocking_sync... Which is still confusing...

Reply:
I'm going to run this through our test suite to see what comes out. As I mentioned in a different comment, it looks like this changes makes sense. Also note that PR #60411 introduces fio tests with nfs-gnesha (with configurations that are default for that, i.e., disabled object cacher).